File: tools/autopilot.mjs
/**
 * Headless autopilot. Drives each generated circuit with a simple but
 * competent racing driver so we can check that every level is actually
 * completable, and measure how much fuel a clean run consumes.
 *
 * Run: node tools/autopilot.mjs [levelCount]
 *
 * The driver is deliberately not perfect — it follows the centreline rather
 * than a racing line — so the fuel it burns is a fair "decent human" budget.
 */
import { Vehicle, SPEC } from '../src/physics.js';
import { Track, levelConfig } from '../src/track.js';

const H = 1 / 120;

/** Pure-pursuit steering plus a curvature-derived speed target. */
function drive(v, track, hint) {
  const loc = track.locate(v.x, v.z, hint);
  const n = track.frames.length - 1;

  // Look ahead a distance that scales with speed.
  const lookAhead = 8 + v.speed * 0.55;
  const step = track.length / n;
  const aheadIdx = (loc.index + Math.round(lookAhead / step)) % n;
  const target = track.frames[aheadIdx].pos;

  // Steering: angle to the aim point, in the car's frame.
  const dx = target.x - v.x, dz = target.z - v.z;
  const local = Math.atan2(dx, dz) - v.yaw;
  const err = Math.atan2(Math.sin(local), Math.cos(local));
  const steer = Math.max(-1, Math.min(1, err / (v.steerLimit || SPEC.maxSteer) * 0.9));

  // Speed target: worst curvature over the next few hundred metres sets the
  // braking point, via v = sqrt(a_lat_max / |k|).
  let worst = 0;
  const scan = Math.round((30 + v.speed * 2.2) / step);
  for (let i = 0; i < scan; i++) {
    worst = Math.max(worst, Math.abs(track.curvature[(loc.index + i) % n]));
  }
  const q = 0.5 * SPEC.airDensity * SPEC.frontalArea * v.speed * v.speed;
  const aLat = SPEC.tyre.rear.D * (SPEC.mass * 9.81 + q * SPEC.liftCoefficient) / SPEC.mass;
  const vTarget = worst > 1e-5 ? Math.sqrt(aLat / worst) : 110;

  // Keep it near the centreline if it has drifted wide.
  const correction = Math.max(-0.35, Math.min(0.35, -loc.lateral * 0.06));

  const dv = vTarget - v.speed;
  return {
    steer: Math.max(-1, Math.min(1, steer + correction)),
    throttle: dv > 1 ? Math.min(1, dv / 6) : 0,
    brake: dv < -1.5 ? Math.min(1, -dv / 12) : 0,
    hint: loc.index
  };
}

export function simulate(levelIndex, { maxSeconds = 400, fuel = 1e6 } = {}) {
  const cfg = levelConfig(levelIndex);
  const track = new Track(cfg);
  const v = new Vehicle();
  const pose = track.startPose;
  v.reset(pose.x, pose.z, pose.yaw);
  v.fuel = fuel;

  let hint = 0, t = 0, next = 0, hits = 0, offTrackTime = 0, topSpeed = 0;
  while (t < maxSeconds && next < track.checkpoints.length) {
    const input = drive(v, track, hint);
    hint = input.hint;

    const surf = track.surface(v.x, v.z, hint);
    v.gripScale = surf.grip;
    if (surf.offTrack) offTrackTime += H;
    v.step(H, input);
    if (surf.hit) { v.collide(surf.hit.nx, surf.hit.nz, surf.hit.penetration + 0.02); hits++; }

    const cp = track.checkpoints[next];
    const loc = track.locate(v.x, v.z, hint);
    if (Math.abs(loc.index - cp.index) < 12) next++;

    topSpeed = Math.max(topSpeed, v.speed);
    t += H;
    if (v.fuel <= 0) break;
  }

  return {
    name: cfg.name, cfg, track,
    finished: next >= track.checkpoints.length,
    time: t, fuel: v.fuelUsed, hits, offTrackTime, topSpeed,
    trackLength: track.length,
    tightestRadius: track.tightestRadius,
    avgSpeed: track.length / t
  };
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const count = Number(process.argv[2] || 6);
  console.log('lvl  track                 len   time    avg     top    fuel    hits  off   minR');
  for (let i = 0; i < count; i++) {
    const r = simulate(i);
    console.log(
      String(i + 1).padEnd(4),
      r.name.padEnd(21),
      `${Math.round(r.trackLength)}m`.padStart(5),
      (r.finished ? `${r.time.toFixed(1)}s` : 'DNF').padStart(7),
      `${Math.round(r.avgSpeed * 3.6)}`.padStart(5),
      `${Math.round(r.topSpeed * 3.6)}`.padStart(6),
      `${r.fuel.toFixed(2)}kg`.padStart(8),
      String(r.hits).padStart(5),
      `${r.offTrackTime.toFixed(1)}s`.padStart(6),
      `R${Math.round(r.tightestRadius)}m`.padStart(7)
    );
  }
}

File: core/src/main/kotlin/dev/racer/core/Autopilot.kt
package dev.racer.core

import kotlin.math.abs
import kotlin.math.atan
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan
import kotlin.math.roundToInt

/**
 * A headless racing driver.
 *
 * Used by the test suite to prove every generated circuit is actually
 * completable, and to measure how much fuel a clean run consumes — which is
 * where the levels' fuel budgets come from.
 *
 * It follows the centreline rather than a racing line, so it is a fair "decent
 * human" reference rather than a perfect lap.
 */
object Autopilot {

    class Result(
        val cfg: LevelConfig,
        val track: Track,
        val finished: Boolean,
        val time: Double,
        val fuelUsed: Double,
        val wallHits: Int,
        val offTrackTime: Double,
        val topSpeed: Double
    ) {
        val avgSpeed: Double get() = if (time > 0) track.length / time else 0.0
    }

    private const val H = 1.0 / 120.0

    /** How far ahead to look for corners when picking a braking point (m). */
    private const val LOOKAHEAD_METRES = 350.0

    /**
     * Deceleration assumed available for braking (m/s^2).
     *
     * Deliberately pessimistic: it must hold at the *end* of the braking zone,
     * where the car is slow and has almost no downforce, so the usable figure
     * is roughly the mechanical grip limit (1.85 g) rather than the much larger
     * number available at 300 km/h. Assuming the high-speed value makes the
     * driver brake late and arrive at the apex too fast.
     */
    var brakeDecel = 8.0

    private const val TOP_SPEED_TARGET = 110.0

    /** Take corners under the theoretical limit, like a human. */
    var cornerSpeedMargin = 0.76

    // Exposed as vars so the tuning sweep in the tests can search them.
    var headingGain = 0.8
    var crossTrackGain = 0.9
    var crossTrackSoften = 8.0

    /** How fast the driver may move the steering command, per second. */
    private const val STEER_RATE = 3.0

    /**
     * One control step: Stanley steering plus a braking-distance speed target.
     */
    fun input(v: Vehicle, track: Track, hint: Int, previousSteer: Double = 0.0, dt: Double = H): Pair<Input, Int> {
        val loc = track.locate(v.x, v.z, hint)
        val n = track.frames.size - 1
        val step = track.length / n

        // --- how much grip is available right now ----------------------------
        val q = 0.5 * Spec.AIR_DENSITY * Spec.FRONTAL_AREA * v.speed * v.speed
        val aLat = Spec.TYRE_REAR.d * (Spec.MASS * 9.81 + q * Spec.LIFT_COEFFICIENT) / Spec.MASS

        // --- steering ---------------------------------------------------------
        // Stanley control: the lock the corner itself needs (feedforward), plus
        // corrections on heading error and on cross-track error, the latter
        // softened by speed so it is gentle at 300 km/h and firm at walking pace.
        val steerLimit = if (v.steerLimit > 1e-6) v.steerLimit else Spec.MAX_STEER

        // Look a little ahead so the wheel is already turning as the corner
        // arrives rather than after it has started.
        val leadFrames = ((2.0 + v.speed * 0.25) / step).roundToInt()
        val kAhead = track.curvature[(loc.index + leadFrames) % n]

        // Curvature is measured as -dYaw/ds (see Track.curvatures), so the lock
        // the corner requires is atan(wheelbase * -k).
        val feedForward = atan(-(Spec.A + Spec.B) * kAhead)

        val trackYaw = atan2(loc.frame.tangent.x, loc.frame.tangent.z)
        val headingError = wrapPi(trackYaw - v.yaw)
        // Car right of the centreline (lateral > 0) needs left lock.
        val crossTrack = -atan(crossTrackGain * loc.lateral / (v.speed + crossTrackSoften))

        val delta = feedForward + headingGain * headingError + crossTrack

        // The physics maps full input to the grip-limited steering angle, which
        // at 240 km/h is about two degrees — so a small angle error saturates
        // the input. Rate limiting the command stops that saturation arriving
        // as an instant lock-to-lock flick, which snaps the rear loose.
        val wanted = clamp(delta / steerLimit, -1.0, 1.0)
        val steer = previousSteer + clamp(wanted - previousSteer, -STEER_RATE * dt, STEER_RATE * dt)

        // --- speed ------------------------------------------------------------
        // The speed allowed *here* is whatever can still be shed in time to
        // arrive at each corner ahead at its own limit:
        //   v <= sqrt(v_corner^2 + 2 * a_brake * distance)
        // Taking the minimum over the lookahead gives a real braking point.
        val scan = (LOOKAHEAD_METRES / step).roundToInt()
        var vTarget = TOP_SPEED_TARGET
        for (i in 0 until scan) {
            val k = abs(track.curvature[(loc.index + i) % n])
            if (k < 1e-5) continue
            val vCorner = cornerSpeed(k) * cornerSpeedMargin
            val allowed = sqrt(vCorner * vCorner + 2 * brakeDecel * (i * step))
            vTarget = min(vTarget, allowed)
        }

        val dv = vTarget - v.speed
        return Input(
            throttle = if (dv > 1) min(1.0, dv / 6) else 0.0,
            brake = if (dv < -1.5) min(1.0, -dv / 12) else 0.0,
            steer = steer
        ) to loc.index
    }


    /**
     * The fastest a corner of curvature [k] can be taken, solved
     * self-consistently for downforce.
     *
     * Grip depends on speed (downforce rises with v^2) and so does the demand
     * (v^2 * k), so the corner speed is the fixed point of
     *     v^2 * k = D * (W + c * v^2) / m
     * which rearranges to v^2 = D * W / (m * k - D * c).
     *
     * Evaluating grip at the *current* speed instead — which is much higher on
     * the approach — overestimates the corner speed badly and puts the car in
     * the barrier. If the denominator is non-positive, downforce grows faster
     * than the demand and the corner is flat out.
     */
    fun cornerSpeed(k: Double): Double {
        val c = 0.5 * Spec.AIR_DENSITY * Spec.FRONTAL_AREA * Spec.LIFT_COEFFICIENT
        val denominator = Spec.MASS * k - Spec.TYRE_REAR.d * c
        if (denominator <= 1e-6) return TOP_SPEED_TARGET
        val vSquared = Spec.TYRE_REAR.d * Spec.MASS * 9.81 / denominator
        return min(TOP_SPEED_TARGET, sqrt(vSquared))
    }

    private fun wrapPi(a: Double) = atan2(sin(a), cos(a))

    fun simulate(levelIndex: Int, maxSeconds: Double = 400.0, fuel: Double = 1e6): Result {
        val cfg = Levels.config(levelIndex)
        val track = Track(cfg)
        val v = Vehicle()
        val (sx, sz, syaw) = track.startPose
        v.reset(sx, sz, syaw)
        v.fuel = fuel

        var hint = 0
        var steer = 0.0
        var t = 0.0
        var next = 0
        var hits = 0
        var offTrackTime = 0.0
        var topSpeed = 0.0

        while (t < maxSeconds && next < track.checkpoints.size) {
            val (inp, newHint) = input(v, track, hint, steer, H)
            hint = newHint
            steer = inp.steer

            val surf = track.surface(v.x, v.z, hint)
            v.gripScale = surf.grip
            if (surf.offTrack) offTrackTime += H
            v.step(H, inp)
            surf.hit?.let { v.collide(it.nx, it.nz, it.penetration + 0.02); hits++ }

            val loc = track.locate(v.x, v.z, hint)
            val n = track.frames.size - 1
            if (abs(wrapIndex(loc.index - track.checkpoints[next], n)) < 12) next++

            topSpeed = max(topSpeed, v.speed)
            t += H
            if (v.fuel <= 0) break
        }

        return Result(cfg, track, next >= track.checkpoints.size, t, v.fuelUsed, hits, offTrackTime, topSpeed)
    }

    /**
     * Compare frame indices the short way round the loop. The finish line sits
     * at index n, which is the same place as index 0, so a plain subtraction
     * would read the finish as a whole lap away just as you cross it.
     */
    fun wrapIndex(d: Int, n: Int): Int {
        var r = d % n
        if (r > n / 2) r -= n
        if (r < -n / 2) r += n
        return r
    }
}
